### PR TITLE
Add New Undistortion Models

### PR DESCRIPTION
**Pull request by servantftransperfect**
Remove all *distortion* models related to 3DE
Add them as new undistortion models (With corrected equations)
Export those new undistortion models to 3DE nuke nodes
Add a pixelaspectratio property in the undistortion class (and upgrade associated I/O)

Related Meshroom PR: https://github.com/alicevision/Meshroom/pull/2446

### CODE REVIEW

File: src/aliceVision/camera/Undistortion3DEClassicLD.cpp [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    d_intermediate_d_params(5, 3) = 1.0;
+    d_intermediate_d_params(6, 0) = 1.0;
+    d_intermediate_d_params(7, 4) = 1.0;
+    d_intermediate_d_params(8, 4) = 2.0;
+    d_intermediate_d_params(9, 4) = 1.0;
+
+    
+
+    double x = p.x();
+	double y = p.y();
```
Comment (fabiencastan) [inline src/aliceVision/camera/Undistortion3DEClassicLD.cpp]: remove all tabs

File: src/aliceVision/camera/Undistortion3DEA4.cpp [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+    const double& cy44 = _undistortionParams[9];
+
+    const double& xu = p.x();
+    const double& yu = p.y();
+
+    double r2 = xu * xu + yu * yu;
+    double r4 = r2 * r2;
+    double theta = std::atan2(yu, xu);
+    double cos_2theta = cos(2.0 * theta);
+    double cos_4theta = cos(4.0 * theta);
```
Comment (simogasp) [inline src/aliceVision/camera/Undistortion3DEA4.cpp]: const

File: src/aliceVision/camera/Undistortion3DEA4.cpp [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+    double d_xxxx_d_x = 4.0 * xu * xx; 
+    double d_xxyy_d_x = 2.0 * xu * yy; 
+    double d_yy_d_y = 2.0 * yu;
+    double d_yyyy_d_y = 4.0 * yu * yy;
+    double d_xxyy_d_y = 2.0 * yu * xx;
+
+    double d_distx_d_x = cx_xx * d_xx_d_x + cx_xxxx * d_xxxx_d_x + cx_xxyy * d_xxyy_d_x; 
+    double d_distx_d_y = cx_yy * d_yy_d_y + cx_xxyy * d_xxyy_d_y + cx_yyyy * d_yyyy_d_y;
+    double d_disty_d_x = cy_xx * d_xx_d_x + cy_xxxx * d_xxxx_d_x + cy_xxyy * d_xxyy_d_x; 
+    double d_disty_d_y = cy_yy * d_yy_d_y + cy_xxyy * d_xxyy_d_y + cy_yyyy * d_yyyy_d_y;
```
Comment (simogasp) [inline src/aliceVision/camera/Undistortion3DEA4.cpp]: const

File: src/aliceVision/camera/Undistortion3DEA4.cpp [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+    const double& cy44 = _undistortionParams[9];
+
+    const double& xu = p.x();
+    const double& yu = p.y();
+
+    double r2 = xu * xu + yu * yu;
+    double r4 = r2 * r2;
+    double theta = std::atan2(yu, xu);
+    double cos_2theta = cos(2.0 * theta);
+    double cos_4theta = cos(4.0 * theta);
```
Comment (simogasp) [inline src/aliceVision/camera/Undistortion3DEA4.cpp]: const

File: src/aliceVision/camera/Undistortion3DEClassicLD.cpp [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    double cyxx = c4;
+    double cyyx = 2.0 * c4;
+    double cyyy = c4;
+
+    double x = p.x();
+    double y = p.y();
+    double x2 = x * x;
+    double y2 = y * y;
+    double x4 = x2 * x2;
+    double y4 = y2 * y2;
```
Comment (simogasp) [inline src/aliceVision/camera/Undistortion3DEClassicLD.cpp]: const

File: src/aliceVision/camera/Undistortion3DEClassicLD.cpp [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+    float x2y2 = x2 * y2;
```
Comment (simogasp) [inline src/aliceVision/camera/Undistortion3DEClassicLD.cpp]: any particular reason to a have a float while the result is a product of doubles?

File: src/aliceVision/camera/Undistortion3DEClassicLD.cpp [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+
+    double x = p.x();
+    double y = p.y();
+    double x2 = x * x;
+    double y2 = y * y;
+    double x3 = x2 * x;
+    double y3 = y2 * y;
+    double x4 = x2 * x2;
+    double y4 = y2 * y2;
+    double x2y2 = x2 * y2;
```
Comment (simogasp) [inline src/aliceVision/camera/Undistortion3DEClassicLD.cpp]: const

File: src/aliceVision/camera/Undistortion3DEClassicLD.cpp [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+    double d2 = (1.0 + cyx * x2 + cyy * y2 + cyxx * x4 + cyyx * x2y2 + cyyy * y4);
+    
+    np.x() = x * d1;
+    np.y() = y * d2;
+
+    double d_d1_d_x = cxx * 2.0 * x + cxxx * 4.0 * x3 + cxxy * 2.0 * x * y2;
+    double d_d1_d_y = cxy * 2.0 * y + cxxy * 2.0 * y * x2 + cxyy * 4.0 * y3;
+
+    double d_d2_d_x = cyx * 2.0 * x + cyxx * 4.0 * x3 + cyyx * 2.0 * x * y2;
+    double d_d2_d_y = cyy * 2.0 * y + cyyx * 2.0 * y * x2 + cyyy * 4.0 * y3;
```
Comment (simogasp) [inline src/aliceVision/camera/Undistortion3DEClassicLD.cpp]: const

File: src/aliceVision/camera/Undistortion3DEClassicLD.cpp [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+    double x3 = x2 * x;
+    double y3 = y2 * y;
+    double x4 = x2 * x2;
+    double y4 = y2 * y2;
+    double x2y2 = x2 * y2;
+
+    Vec2 np;
+
+    double d1 = (1.0 + cxx * x2 + cxy * y2 + cxxx * x4 + cxxy * x2y2 + cxyy * y4);
+    double d2 = (1.0 + cyx * x2 + cyy * y2 + cyxx * x4 + cyyx * x2y2 + cyyy * y4);
```
Comment (simogasp) [inline src/aliceVision/camera/Undistortion3DEClassicLD.cpp]: const

File: src/aliceVision/camera/Undistortion3DEClassicLD.cpp [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+    double cxy = (c2 + cx) / sq;
+    double cxxx = c4 / sq;
+    double cxxy = 2.0 * c4 / sq;
+    double cxyy = c4 / sq;
+    
+    double cyx = c2 + cy;
+    double cyy = c2;
+    double cyxx = c4;
+    double cyyx = 2.0 * c4;
+    double cyyy = c4;
```
Comment (simogasp) [inline src/aliceVision/camera/Undistortion3DEClassicLD.cpp]: const

File: src/aliceVision/camera/Undistortion3DEClassicLD.cpp [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+
+    double x = p.x();
+    double y = p.y();
+    double x2 = x * x;
+    double y2 = y * y;
+    double x3 = x2 * x;
+    double y3 = y2 * y;
+    double x4 = x2 * x2;
+    double y4 = y2 * y2;
+    double x2y2 = x2 * y2;
```
Comment (simogasp) [inline src/aliceVision/camera/Undistortion3DEClassicLD.cpp]: const

File: src/aliceVision/camera/Undistortion3DERadial4.cpp [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+    const double & b_cylindric_bending = _undistortionParams[7];
+
+    const double q = sqrt(1.0 + b_cylindric_bending);
+    const double cphi = cos(phi_cylindric_direction);
+    const double sphi = sin(phi_cylindric_direction);
+
+    double m11 = cphi * cphi * q + sphi * sphi / q;
+    double m12 = (q - 1.0 / q) * cphi * sphi;
+    double m21 = (q - 1.0 / q) * cphi * sphi;
+    double m22 = cphi * cphi / q + sphi * sphi * q;
```
Comment (simogasp) [inline src/aliceVision/camera/Undistortion3DERadial4.cpp]: const

File: src/aliceVision/camera/Undistortion3DERadial4.cpp [hunk 13]
```diff
@@ -0,0 +1,267 @@
...
+
+    double d_m11_d_sphi = 2.0 * sphi / q;
+    double d_m12_d_sphi = (q - 1.0 / q) * cphi;
+    double d_m21_d_sphi = (q - 1.0 / q) * cphi;
+    double d_m22_d_sphi = 2.0 * sphi * q;
+
+    double d_m11_d_q = cphi * cphi + sphi * sphi * d_invq_d_q;
+    double d_m12_d_q = (1.0 - d_invq_d_q) * cphi * sphi;
+    double d_m21_d_q = (q - d_invq_d_q) * cphi * sphi;
+    double d_m22_d_q = cphi * cphi * d_invq_d_q + sphi * sphi;
```
Comment (simogasp) [inline src/aliceVision/camera/Undistortion3DERadial4.cpp]: const

File: src/aliceVision/camera/Undistortion3DERadial4.cpp [hunk 14]
```diff
@@ -0,0 +1,267 @@
...
+    const double & phi_cylindric_direction = _undistortionParams[6];
+    const double & b_cylindric_bending = _undistortionParams[7];
+
+    const double q = sqrt(1.0 + b_cylindric_bending);
+    const double cphi = cos(phi_cylindric_direction);
+    const double sphi = sin(phi_cylindric_direction);
+
+    double d_q_b_cylindric_bending = 1.0 / (2.0 * q);
+    double d_cphi_d_phi = -sphi;
+    double d_sphi_d_phi = cphi;
```
Comment (simogasp) [inline src/aliceVision/camera/Undistortion3DERadial4.cpp]: const

**Review by simogasp — DISMISSED** [review]

File: src/aliceVision/camera/Undistortion3DEClassicLD.cpp (hunk 6 above)
Comment (servantftransperfect) [reply src/aliceVision/camera/Undistortion3DEClassicLD.cpp]: Thanks.

**Review by cbentejac — APPROVED** [review]

**Pull request merged**